### PR TITLE
Added function preconditions to initial function CFG edge

### PR DESCRIPTION
**Pull request by michaelZhaoCode**
## Motivation and Context

The CFG visitor creates a new CFG for any function blocks encountered. Assuming we are checking preconditions, these blocks would only be ran if the preconditions are satisfied. Thus, the initial edge in these function CFGs now stores the preconditions of the function, as an `and` node in the condition attribute of the edge. 

## Your Changes

Created helper function to extract preconditions, combine them into an `and` expression, and parse into AST node
Used helper function to set CFGEdge condition in the initial function CFG edge to store the condition node
Allowed `parse_arguments` from contracts to take in AST function definition nodes with no docstring
Created tests to verify conditions were set properly

**Description**:

**Type of change** (select all that apply):

- [x] New feature (non-breaking change which adds functionality)

## Testing

Implemented tests to ensure condition was set properly.

## Questions and Comments (if applicable)

Should this be added to the changelog file?

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the CI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls. 
- [x] I have added tests for my changes. 
- [x] I have updated the CHANGELOG.md file.

### CODE REVIEW

File: python_ta/cfg/visitor.py [hunk 1]
```diff
@@ -435,3 +441,16 @@ def _get_raise_exc(node: nodes.Raise) -> str:
...
+
+def _get_preconditions_node(func: nodes.FunctionDef) -> Optional[nodes.BoolOp]:
+    """A helper method that takes in a function definition node, retrieves its preconditions, and then parses them
+    into a AST node representing all the preconditions combined in an and statement. Returns None if there are no
+    preconditions."""
+    assertions = parse_assertions(func)
+    if not assertions:
+        return None
+    precondition_string = " and ".join(assertions)
+    condition, *_ = parse(precondition_string).nodes_of_class(nodes.Expr)
```
Comment (david-yz-liu) [inline python_ta/cfg/visitor.py:454]: Two comments here:

1. You should be able to use `extract_node` to obtain the minimal expression node, rather than using `nodes_of_class`
2. More importantly, this will fail if users specify non-Python preconditions (e.g. if there's a typo, or an English precondition). You should add test cases for this.

    In terms of expected behaviour, the easier approach is to return `None` in this case. The more complex approach is to filter out the non-Python preconditions and only join together the Python preconditions. I'll leave it to you to decide which approach to take, based on how much time you'd like to commit to this issue. 😃

File: python_ta/cfg/visitor.py [hunk 2]
```diff
@@ -435,3 +441,16 @@ def _get_raise_exc(node: nodes.Raise) -> str:
...
+    x = 5
```
Comment (david-yz-liu) [inline python_ta/cfg/visitor.py:455]: this line should be removed

File: python_ta/cfg/visitor.py [hunk 3]
```diff
@@ -435,3 +442,28 @@ def _get_raise_exc(node: nodes.Raise) -> str:
...
+    """A helper method that takes in a function definition node, retrieves its preconditions, and then parses them
+    into a AST node representing all the preconditions combined in an and statement. Returns None if there are no
+    preconditions."""
+    assertions = parse_assertions(func)
+    valid_assertions = [assertion for assertion in assertions if _is_python_precondition(assertion)]
+    # for now, return None if at least one non-valid assertion
+    # remove the length comparison and improve valid precondition checker to change this behaviour
+    if not valid_assertions or len(valid_assertions) != len(assertions):
+        return None
+    precondition_string = " and ".join(valid_assertions)
```
Comment (david-yz-liu) [inline python_ta/cfg/visitor.py:457]: One other thing, you should really wrap each assertion in parentheses, to avoid any issues regarding `and` and [operator precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence). It's the same issue as C macros that you learned about in CSC209.

**Review by david-yz-liu — COMMENTED** [review]
@michaelZhaoCode good work. I left one comment; also, I think this change warrants an entry under the Changelog "enhancements".

**Review by david-yz-liu — APPROVED** [review]
